File: js/kmeans.js
var fs = require('fs');

var pointsFile = process.argv[2];
var seedsFile = process.argv[3];

function zip(lhs, rhs) {
  return lhs.map(function(l,i) { return [l, rhs[i]]; });
}

function Vec(pts) {
  this.pts = pts;
}

Vec.prototype.dot = function(that) {
  return zip(this.pts, that.pts)
    .map(function (x) { return x[0] * x[1]; })
    .reduce(function (a, b) { return a + b; });
}

Vec.prototype.add = function(that) {
  return new Vec(
    zip(this.pts, that.pts)
    .map(function (x) { return x[0] + x[1]; }));
}

Vec.prototype.sub = function(that) {
  return new Vec(
    zip(this.pts, that.pts)
    .map(function (x) { return x[0] - x[1]; }));
}

Vec.prototype.mult = function(a) {
  return new Vec(this.pts.map(function (e) { return e * a; }));
}

Vec.prototype.norm = function() {
  return this.dot(this);
}

function parsePoint(s) {
  return new Vec(s.trim().split(" ").map(parseFloat));
}

function bootstrap(pointsData, seedsData) {
  var points = pointsData.trim().split("\n").map(parsePoint);
  var seeds = seedsData.trim().split("\n").map(parsePoint);
  kmeans(points, seeds, 20).forEach(function (e) { console.log(e); });
}

function argmin(elems) {
  var minSoFar = elems[0];
  var minIdxSoFar = 0;
  for (var i = 1; i < elems.length; i++) {
    if (elems[i] < minSoFar) {
      minSoFar = elems[i];
      minIdxSoFar = i;
    }
  }
  return minIdxSoFar;
}

function which(pt, clusters) {
  return argmin(clusters.map(function (c) { return pt.sub(c).norm(); }));
}

function centroid(pts) {
  return pts
    .reduce(function (a, b) { return a.add(b); })
    .mult(1. / pts.length);
}

function kmeans(points, seeds, iters) {
  var clusters = seeds.slice(0); // clone
  for (var i = 0; i < iters; i++) {
    var assignments = [];
    for (var j = 0; j < clusters.length; j++) {
      assignments.push([]);
    }
    points.forEach(function (pt) {
      assignments[which(pt, clusters)].push(pt);
    });
    clusters = zip(assignments, clusters)
      .map(function (e) { return e[0].length === 0 ? e[1] : centroid(e[0]); });
  }
  return clusters
}

fs.readFile(pointsFile, 'ascii', function (err, pointsData) {
  fs.readFile(seedsFile, 'ascii', function (err, seedsData) {
    bootstrap(pointsData, seedsData);
  });
});

File: py/kmeans.py
import math
import sys
import itertools as it


class Vec(object):
    def __init__(self, pts):
        self.pts = pts

    def dot(self, that):
        return sum(x * y for x, y in zip(self.pts, that.pts))

    def norm(self):
        return math.sqrt(self.dot(self))

    def __add__(self, that):
        return Vec([x + y for x, y in zip(self.pts, that.pts)])

    def __sub__(self, that):
        return Vec([x - y for x, y in zip(self.pts, that.pts)])

    def __mul__(self, that):
        """Scalar multiplication

        """
        return Vec([that * x for x in self.pts])

    def __str__(self):
        return str(self.pts)

    def __repr__(self):
        return repr(self.pts)


def argmin(elems):
    minSoFar, minIdxSoFar = elems[0], 0
    for i, e in it.islice(enumerate(elems), 1, None):
        if e < minSoFar:
            minSoFar, minIdxSoFar = e, i
    return minIdxSoFar


def kmeans(points, seeds, iters):

    def centroid(pts):
        return reduce(lambda x, y: x + y, pts) * (1. / len(pts))

    clusters = list(seeds)
    for t in xrange(iters):
        assignments = [[] for _ in clusters]
        for pt in points:
            dists = map(lambda c: (c - pt).norm(), clusters)
            assignments[argmin(dists)].append(pt)
        clusters = [c if not elems else centroid(elems)
                    for c, elems in zip(clusters, assignments)]
    return clusters


def parse(fname):
    with open(fname, 'r') as fp:
        for l in fp:
            yield Vec(map(float, l.strip().split()))


def main():
    pointsFile, seedsFile = sys.argv[1:]
    points = list(parse(pointsFile))
    seeds = list(parse(seedsFile))
    clusters = kmeans(points, seeds, 20)
    for c in clusters:
        print c


if __name__ == '__main__':
    main()
